### PR TITLE
[Pagination] Feature 8609 : new features

### PR DESCRIPTION
**Pull request by NicolasEYSSERIC**
- adding current items area and custom number of items per page (20 - 50 - 100 - All)
- rationalizing implementation about pagination (deleting PaginationSilverpeasV5.java, changing PaginationTag.java for using PaginationSP)
- removing useless code on tag 'pagination'
- homogenizing look n' feel of Java pagination with 'pagination' Javascript plugin
- using new pagination features on directory and search engine results page

Please, do not forget PR on Silverpeas-Components

### CODE REVIEW

File: core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java [hunk 1]
```diff
@@ -114,29 +128,56 @@ public String printIndex(String javascriptFunc) {
...
+        if (getNbItemsPerPage() == specialValue) {
+          result.append("<a class=\"selected\">").append(getString("GEF.pagination.all")).append("</a>");
+        } else {
+          result.append(
+              "<a href=\"").append(getNbItemsPerPageLink(javascriptFunc, specialValue)).append("\" title=\"").append(getString("GEF.pagination.all.title")).append("\">").append(getString("GEF.pagination.all")).append("</a>");
+        }
+        result.append("</div>");
+      }
+
+      boolean displayJumper = nbPage > 12;
```
Comment (silveruser) [inline core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java]: MINOR Assign this magic number 12 to a well-named constant, and use the constant instead. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)

File: core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java [hunk 2]
```diff
@@ -46,66 +46,80 @@ public String printCounter() {
...
+        for (int i = 1; i <= nbPage; i++) {
+          if (i == currentPage) {
+            result.append("<a class=\"pageOn\">");
             result.append(i).append("</a>");
-            result.append("</div>");
+          } else {
+            int nbDisplayPages = getNumberOfPagesAround();
+            // display 10 pages (or less) before current page
+            // display 10 pages (or less) after current page
+            if (currentPage - nbDisplayPages <= i && i <= currentPage + nbDisplayPages) {
```
Comment (silveruser) [inline core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java]: MINOR Refactor this code to not nest more than 4 if/for/while/switch/try statements. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS134)

File: core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java [hunk 3]
```diff
@@ -114,29 +128,56 @@ public String printIndex(String javascriptFunc) {
...
-              .append(getString("GEF.pagination.lastPage")).append("\"/></a> ");
-          result.append("</div>");
+              "/arrows/arrowDoubleRight.png\" border=\"0\" align=\"absmiddle\" alt=\"")
+              .append(getString("GEF.pagination.lastPage")).append("\"/>");
+          result.append("</a>");
         }
       }
+      result.append("</div>");
 
+      if (nbItemsPerPage && getNbItems() > 25) {
```
Comment (silveruser) [inline core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java]: MINOR Assign this magic number 25 to a well-named constant, and use the constant instead. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)

File: core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java [hunk 4]
```diff
@@ -114,29 +128,56 @@ public String printIndex(String javascriptFunc) {
...
+            } else {
+              result.append("<a href=\"").append(getNbItemsPerPageLink(javascriptFunc, value))
+                  .append("\" title=\"")
+                  .append(getStringWithParam("GEF.pagination.perPage", String.valueOf(value))).append("\">")
+                  .append(value).append("</a>");
+            }
+          }
+        }
+        // add special feature : All
+        int specialValue = 100000;
```
Comment (silveruser) [inline core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java]: MINOR Assign this magic number 100000 to a well-named constant, and use the constant instead. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)

File: core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java [hunk 5]
```diff
@@ -114,29 +128,56 @@ public String printIndex(String javascriptFunc) {
...
+              "/arrows/arrowDoubleRight.png\" border=\"0\" align=\"absmiddle\" alt=\"")
+              .append(getString("GEF.pagination.lastPage")).append("\"/>");
+          result.append("</a>");
         }
       }
+      result.append("</div>");
 
+      if (nbItemsPerPage && getNbItems() > 25) {
+        result.append("<div class=\"pageIndex numberPerPage\">");
+        int[] values = new int[]{25, 50, 100};
```
Comment (silveruser) [inline core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java]: MINOR Assign this magic number 25 to a well-named constant, and use the constant instead. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)
MINOR Assign this magic number 50 to a well-named constant, and use the constant instead. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)
MINOR Assign this magic number 100 to a well-named constant, and use the constant instead. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)

File: core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java [hunk 6]
```diff
@@ -46,66 +46,80 @@ public String printCounter() {
...
   public String printIndex(String javascriptFunc) {
+    return printIndex(javascriptFunc, StringUtil.isNotDefined(javascriptFunc));
+  }
+
+  @Override
+  public String printIndex(String javascriptFunc, boolean nbItemsPerPage) {
     StringBuilder result = new StringBuilder();
 
-    if (getNbItems() > 0 && getNbItems() > getNbItemsPerPage()) {
+    if (getNbItems() > 5) {
```
Comment (silveruser) [inline core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java]: MINOR Assign this magic number 5 to a well-named constant, and use the constant instead. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)

File: core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java [hunk 7]
```diff
@@ -114,29 +128,56 @@ public String printIndex(String javascriptFunc) {
...
       }
+      result.append("</div>");
 
+      if (nbItemsPerPage && getNbItems() > 25) {
+        result.append("<div class=\"pageIndex numberPerPage\">");
+        int[] values = new int[]{25, 50, 100};
+        for (int i=0; i<values.length; i++) {
+          int value = values[i];
+          if (getNbItems() > value || (i != 0 && getNbItems() > values[i-1])) {
+            if (getNbItemsPerPage() == value) {
```
Comment (silveruser) [inline core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java]: MINOR Refactor this code to not nest more than 4 if/for/while/switch/try statements. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS134)

File: core-war/src/main/java/org/silverpeas/web/pdc/control/PdcSearchSessionController.java [hunk 8]
```diff
@@ -118,7 +118,7 @@
   // pagination de la liste des resultats (search Engine)
   private int indexOfFirstResultToDisplay = 0;
   private String displayParamChoices = null; // All || Res || Req
-  private int nbResToDisplay = -1;
+  private int nbResToDisplay = 10;
```
Comment (silveruser) [inline core-war/src/main/java/org/silverpeas/web/pdc/control/PdcSearchSessionController.java]: MINOR Assign this magic number 10 to a well-named constant, and use the constant instead. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)

File: core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationTag.java [hunk 9]
```diff
@@ -41,11 +41,6 @@
   private String action;
   boolean actionIsJsFunction = false;
   private String pageParam;
```
Comment (silveruser) [inline core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationTag.java]: MAJOR Remove this unused "pageParam" private field. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1068)

File: core-war/src/main/java/org/silverpeas/web/directory/servlets/DirectoryRequestRouter.java [hunk 10]
```diff
@@ -147,6 +147,11 @@ public String getDestination(String function, DirectorySessionController directo
         users = directorySC.getLastListOfUsersCalled();
         destination = doPagination(request, users, directorySC);
 
+      } else if (function.equalsIgnoreCase("ChangeNumberItemsPerPage")) {
```
Comment (silveruser) [inline core-war/src/main/java/org/silverpeas/web/directory/servlets/DirectoryRequestRouter.java]: MAJOR Move the "ChangeNumberItemsPerPage" string literal on the left side of this string comparison. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1132)

**Comment by silveruser** [comment]
SonarQube analysis reported 14 issues
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 10 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractArrayPane.java#L80](https://github.com/Silverpeas/Silverpeas-Core/blob/db50cf6b5add9c8d7312a288e006d7a0bec96bb1/core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/arraypanes/AbstractArrayPane.java#L80): The Cyclomatic Complexity of this method "init" is 11 which is greater than 10 authorized. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [PaginationTag.java#L38](https://github.com/Silverpeas/Silverpeas-Core/blob/db50cf6b5add9c8d7312a288e006d7a0bec96bb1/core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationTag.java#L38): Remove this unused "nbPages" private field. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1068)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java [hunk 11]
```diff
@@ -24,200 +24,280 @@
...
+    }
+
+    return fragment.toString();
+  }
+
+  private String getNbItemsPerPageFragment(boolean nbItemsPerPage, String javascriptFunc) {
+    StringBuilder fragment = new StringBuilder();
+    if (nbItemsPerPage && getNbItems() > NUMBERPERPAGE_THRESHOLD) {
+      fragment.append("<div class=\"pageIndex numberPerPage\">");
+      int[] values = new int[]{25, 50, 100};
```
Comment (silveruser) [inline core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationSP.java]: MINOR Assign this magic number 25 to a well-named constant, and use the constant instead. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)
MINOR Assign this magic number 50 to a well-named constant, and use the constant instead. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)
MINOR Assign this magic number 100 to a well-named constant, and use the constant instead. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)

**Comment by silveruser** [comment]
SonarQube analysis reported 8 issues
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractArrayPane.java#L80](https://github.com/Silverpeas/Silverpeas-Core/blob/b4541b1a9ed2451d1a1fea2f1d70ac12ff3d2b43/core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/arraypanes/AbstractArrayPane.java#L80): The Cyclomatic Complexity of this method "init" is 11 which is greater than 10 authorized. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [PaginationTag.java#L38](https://github.com/Silverpeas/Silverpeas-Core/blob/b4541b1a9ed2451d1a1fea2f1d70ac12ff3d2b43/core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/pagination/PaginationTag.java#L38): Remove this unused "nbPages" private field. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1068)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by SilverYoCha** [comment]
Merged.
Closed manually because of manual git operations performed during integration.

**Pull request closed without merging**